flang/lib/Lower/Allocatable.cpp

```diff
@@ -935,6 +935,14 @@ Fortran::lower::genMutableBoxRead(Fortran::lower::FirOpBuilder &builder,
   return fir::AbstractBox{addr};
 }
 
+mlir::Value Fortran::lower::genIsAllocatedOrAssociatedTest(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,                                const fir::MutableBoxValue &box) {
+  auto addr = MutablePropertyReader(builder, loc, box).readBaseAddress();
+  auto idxTy = builder.getIndexType();
+  auto ptrToInt = builder.createConvert(loc, idxTy, addr);
+  auto c0 = builder.createIntegerConstant(loc, idxTy, 0);
+  return builder.create<mlir::CmpIOp>(loc, mlir::CmpIPredicate::ne, ptrToInt, c0);
```

Comment:
This could be an issue if `index` size and target pointer size ever became not equal for some reason.

It should be possible to get the number of bits from the `llvm::Target` for a pointer. I don't remember if `mlir::IndexType` supports such a query...

Reply:
That's a good point. I used it because we used to cast zero index to get null address, but that was OK even if index is not matching target pointer size.
Anyway, index is indeed only guaranteed to be the machine word size, which does not imply much on the pointer size.
I added a getIntPtrType to the builder that returns `i64` (it seems safe to use that for now). Then, if we keep this (i.e, we do not delegate the issue to fir and codegen), I think we would need to access an `llvm::DataLayout` to get this info. It seems we cannot get that from the triple, so not sure how to get that for now.